src/Storages/ObjectStorageQueue/ObjectStorageQueueIFileMetadata.cpp

```diff
@@ -259,15 +259,33 @@ bool ObjectStorageQueueIFileMetadata::trySetProcessing()
     return success;
 }
 
-ObjectStorageQueueIFileMetadata::SetProcessingResponseIndexes
+std::optional<ObjectStorageQueueIFileMetadata::SetProcessingResponseIndexes>
 ObjectStorageQueueIFileMetadata::prepareSetProcessingRequests(Coordination::Requests & requests)
 {
-    [[maybe_unused]] auto state = file_status->state.load();
-    chassert (!(state == FileStatus::State::Processing
+    std::unique_lock processing_lock(file_status->processing_lock, std::defer_lock);
+    if (!processing_lock.try_lock())
+    {
+        /// This is possible in case on the same server
+        /// there are more than one S3(Azure)Queue table processing the same keeper path.
+        LOG_TEST(log, "File {} is being processed on this server by another table on this server", path);
+        return std::nullopt;
+    }
+
+    auto state = file_status->state.load();
+    if (state == FileStatus::State::Processing
         || state == FileStatus::State::Processed
         || (state == FileStatus::State::Failed
             && file_status->retries
-            && file_status->retries >= max_loading_retries)), fmt::format("Oops something went wrong: {}", state));
+            && file_status->retries >= max_loading_retries))
+    {
+        LOG_TEST(log, "File {} has non-processable state `{}` (retries: {}/{})",
+                 path, state, file_status->retries, max_loading_retries);
+
+        /// This is possible in case on the same server
+        /// there are more than one S3(Azure)Queue table processing the same keeper path.
+        LOG_TEST(log, "File {} is being processed on this server by another table on this server", path);
+        return std::nullopt;
+    }
```

Comment:
Though because we do `file_status->onProcessing` only in `finalizeProcessing`, then we could still miss this check and go into `prepareProcessingRequestsImp`, but this will only result in failed multi request, e.g. not a failed assertion as before this fix. This kind of race is fixable, but I am not sure if it is worth it because probably noone is gone be creating several s3queue tables on the same server processing the same keeper path.. 🤔

Reply:
so assert doesn't trigger for higher processing num thread of a single table?

Reply:
No, it does not, because in this case they use the instance of `ObjectStorageQueueSource::FileIterator`, e.g. all that code with optimizations is executed only by one of the threads of processing_thread_num.

Reply:
okay that is the behaviour I described above and the only thing which makes sense to do

> probably noone is going to be creating several s3queue tables on the same server processing the same keeper path..

if someone does this instead of using higher processing thread num than I don't want to help them... :smile:

Reply:
yes, though we do have a test case with this scenario `test_storage_s3_queue::test_streaming_to_many_views`, so tests were flaky because of it crashing the server because of chassert

Reply:
why do we create it on the same keeper_path there?

Reply:
> why do we create it on the same keeper_path there?


That's the purpose of this test - to check that several local tables on the same server can work with the same keeper path... I added this just in case.